### ISSUE TITLE
What should we throw for implementation specific limits?

### ISSUE THREAD

**Issue by flagxor** [WebAssembly/design#993]
We have a number of implementation specific limits, such as number of imported functions, that are relatively synced between browsers.
https://github.com/v8/v8/blob/master/src/wasm/wasm-limits.h
We currently in practice throw WebAssembly.CompileError for these (or return false for validation).
Would it make sense to throw something else such as RangeError, or add WebAssembly.ImplementationLimitError ?

**Comment by kripken** [comment]
If those limits are not temporary, it would be nice to document them somewhere, including rationale? For example,
````
const size_t kV8MaxWasmMemoryPages = 16384; // = 1 GiB
````
seems like a surprising limit, important asm.js apps exist today that need more memory than that.

**Comment by jfbastien** [comment]
I'd rather have these as per-embedder limits, and would document them as such.
This issue is just about acknowledging that such limits can exist, and documenting what behaviors to expect when faced with these limits.

**Comment by rossberg-chromium** [comment]
The underlying question is whether these should be considered validation failures in respective engines. If they are, then `CompileError` seems adequate, but that also affects what the `validate` method returns (false vs exception or other value).

Pro validation failure:
- Devs only need to deal with one kind of compile failure.
- `validate` will never fail.

Cons validation failure:
- De facto, "validation" becomes an implementation-dependent notion.
- Cannot programmatically distinguish between compiler bugs and resource limits (might be a bummer for e.g. error reporting in a JIT compiler).

While the simplicity of the first option is nice, I don't feel entirely comfortable with making the meaning of "valid module" appear implementation-dependent.

**Comment by MikeHolman** [comment]
I agree it would be more elegant for validate to succeed in such situations, but storage types for implementations are probably the same for both compile and validate, and changing this would in practice result in unnecessary overhead and add extra risk to the parser.

For example, we use uint16 for holding argument counts (this is an implementation detail, but we share this with js and on 32 bit platforms it allows to use 16 bits for argument count and 16 bits for flags in a single register). Making argument count infinite would be required to prevent non-normative output, but is obviously impractical.

Also, validate necessarily requires stack/memory allocation, which can already lead to non-normative results.

**Comment by rossberg-chromium** [comment]
@MikeHolman, ah, I didn't assume we would require `validate` to succeed, only that it results in something else than false. An exception would be adequate (but perhaps annoying for client code?).

**Comment by MikeHolman** [comment]
Ah interesting proposal. I didn't even consider that option. It does feel a bit off that this will throw. From a user's perspective I think returning false would be more useful, but I can see an argument for it.

This current state reminds me a bit of Windows APIs where you have to GetLastError, which feels like a hack because they didn't give you all the info from the start. I wonder if we should take a lesson from that book and return something more meaningful than true/false.

**Comment by rossberg-chromium** [comment]
@MikeHolman, I've wondered about that, too. Just having a binary result is generally kind of unhelpful in the error case. For example, Benjamin's test harness performs another compile after a failed validate, just to get a decent error message. I suspect that other users of explicit validation will have the same problem.

Unfortunately, JS does not provide a nice way to express sum types. But I suppose we could return an object of the form `{ok: bool, error?: string}`, where `error` is present iff `ok` is false.

**Comment by bnjbvr** [comment]
Indeed, not having an error message is an issue. At some point we considered also emitting a warning when with the validation error, but that cannot be caught in content code; it's only useful if you look at the web developers console. Having the error in the return value of Validate would be nice. We could even have the `ok` bool set to `true` **and** an error message in the case of hitting internal limits, but that's probably confusing and error-prone.

**Comment by rossberg-chromium** [comment]
I opened #995 to address that particular issue.

**Comment by lukewagner** [comment]
I think the main consideration for whether to have a separate exception type should be whether we think there is any programmatic use for the separate type.  (Because if we're just talking about developers debugging compiler output, the error string is what they actually want.)

I think our existing assumption is that validation pretty much always succeeds in production; that is part of why we're ok with allowing it to be non-deterministically lazy.  Given this, I don't think anyone will be doing much with `CompileError` to begin with (consider that it could be thrown by `compile` *or* at runtime...) and so a second exception type wouldn't be helpful.  I'm guessing mostly `CompileError`s will end up being caught by a `window.onerror` hook that generally prints the `.message` and/or sends a report back to the server.

**Comment by rossberg-chromium** [comment]
@lukewagner, imagine you are developing a runtime compiler framework on top of Wasm, for others to use. If that ever results in a "real" Wasm compile error then that's a bug clients should report to you. If they merely hit implementation limits on their machine then it's not. So your framework might want to distinguish the two classes of errors and report them to the client dev differently.

**Comment by lukewagner** [comment]
Yeah, I suppose that would be a case where the distinction was programmatically useful.  (But I think a more-useful compiler framework would pick and enforce its own conservative limit that it reported eagerly to the user so that it always generated valid wasm.)

**Comment by rossberg-chromium** [comment]
Yes, but that's still unreliable as long as Wasm doesn't specify conservative limits explicitly. And there are quite a few limits to be aware of.

**Comment by lukewagner** [comment]
Yes, but if the tool's goal was to use its own conservative limits, then an internal limit error would be no different than any other validation error: it'd be a bug to report back to the tool author.

**Comment by jfbastien** [comment]
Was there something that we wanted to add to JS.md to document what kind of errors occur when limits are reached?

**Comment by sunfishcode** [comment]
As normatively documented [here](https://webassembly.github.io/spec/js-api/index.html#limits), in the JS API it is required to throw a `CompileError`.
